### DISCUSSION TITLE
Function to show documentation for thing at point

### DISCUSSION THREAD

**Discussion by rrudakov** [joaotavora/eglot#1034] [Ideas]
Hi! Is it possible to disable showing hover information in the minibuffer and implement a function to show it in a new window?

I have the following problems:
- For java projects with jdtls I only see signatures in the minibuffer
<img width="615" alt="image" src="https://user-images.githubusercontent.com/6093590/190105846-710193c1-50d8-4647-986b-8c5272f48090.png">

- For rust-analyzer I mostly see hover information and minibuffer is being resized every time I move point to another symbol:
<img width="610" alt="image" src="https://user-images.githubusercontent.com/6093590/190106394-33440ffb-908f-4409-9514-b7a6807308ac.png">

I've tried:
- To set `(setq eldoc-documentation-strategy #'eldoc-documentation-compose)` in the `eglot-managed-mode-hook`, but it's still distracting to see hover information when I don't need it.
- To set 
```emacs-lisp
(setq eldoc-documentation-functions
                    (thread-last
                      eldoc-documentation-functions
                      (remove #'eglot-hover-eldoc-function)))
``` 
in the `eglot-managed-mode-hook`, but in this case I don't know how to see hover information when I need to.

P. S. I used lsp-mode before and there were a function `lsp-describe-thing-at-point` to show the documentation in a new window, I think I'm looking for an alternative for this functionality.

**Comment by manuel-uberti** [comment]
I am not sure whether this is the proper way to do it, but if you set `eldoc-display-functions` to something like:
```emacs-lisp
(setq eldoc-display-functions '(eldoc-display-in-buffer))
```
And then do `M-x eldoc-print-current-symbol-info` (or just `M-x eldoc`) you should get the window you're looking for.

Personally, I'd recommend not to change `eldoc-display-functions` globally though.

**Reply by rrudakov** [reply]
Thanks for the suggestion. I've tested it. With this option eldoc doesn't show anything in the minibuffer, but I'd still like to see signatures or flymake errors.

**Reply by manuel-uberti** [reply]
I see, then you need something different. My answer was related to your question:
> Is it possible to disable showing hover information in the minibuffer and implement a function to show it in a new window?

**Reply by rrudakov** [reply]
Yeah, the question was specifically about `hover` information. Eglot sets `eldoc-documentation-functions` to `'(eglot-signature-eldoc-function eglot-hover-eldoc-function flymake-eldoc-function)`, so if signature is available, hover won't be ever shown.

**Reply by joaotavora** [reply]
You still need a function to put in `eldoc-display-functions`, something that is smart enough to display only some types of documentation in a separate window.  There's no such thing built-in, but if you do that sort of thing, be sure to propose it here and then we can consider it for inclusion in Emacs (`eldoc.el`).

You may need Eglot to mark different types of documentation so you can sort through them.  It doesn't do this at the moment.  But this is a second step.  If you're interested in this functionality you should tackle the elements of `eldoc-display-functions` first.

**Reply by rrudakov** [reply]
I was able to achieve the desired behaviour using the following code:
```emacs-lisp
  (add-hook 'eglot-managed-mode-hook
            (lambda ()
              (setq eldoc-documentation-functions
                    (thread-last
                      eldoc-documentation-functions
                      (remove #'eglot-hover-eldoc-function)))))

  (defun rr/eglot-describe-thing-at-point ()
    (interactive)
    (when (eglot--server-capable :hoverProvider)
      (let ((buf (current-buffer)))
        (jsonrpc-async-request
         (eglot--current-server-or-lose)
         :textDocument/hover (eglot--TextDocumentPositionParams)
         :success-fn (eglot--lambda ((Hover) contents range)
                       (eglot--when-buffer-window buf
                         (when-let ((info (unless (seq-empty-p contents)
                                            (eglot--hover-info contents range))))
                           (with-help-window (help-buffer)
                             (with-current-buffer (help-buffer)
                               (insert info))))))
         :deferred :textDocument/hover))))

```

Does it make sense to include it in eglot? IMO, it's more consistent with default emacs behaviour: for example for emacs lisp eldoc shows only signature, and if you need a documentation you should call `describe-*` function.

**Reply by joaotavora** [reply]
> Does it make sense to include it in eglot? 

Not really, in my opinion.  One should use `eldoc.el`'s API so that simply `M-x eldoc` can be used for this. If that API  is insufficient, it can be extended.
